### ISSUE TITLE
Freeze layers during training

### ISSUE THREAD

**Issue by mmartin56** [AlexeyAB/darknet#4133]
Hi @AlexeyAB ,

1) During training, are backbone weights 'frozen' by default? I think the answer is no (i.e. weights in darknet53 are modified) but I'd just like to double check.

2) I understand that using `./darknet partial` and the approach described [here](https://github.com/AlexeyAB/darknet#how-to-improve-object-detection) will freeze all weights in the backbone? I.e. layers 0 to 81.

3) Would it make sense from a ML perspective to push that logic even further and freeze all weights except the yolo layers? That would be layers 0 to 106, except layers 82, 94, and 106.

**Comment by AlexeyAB** [comment]
`stopbackward=1` param freezes weights

> to speedup training (with decreasing detection accuracy) do Fine-Tuning instead of Transfer-Learning, set param stopbackward=1 here: https://github.com/AlexeyAB/darknet/blob/6d44529cf93211c319813c90e0c1adb34426abe5/cfg/yolov3.cfg#L548 then do this command: ./darknet partial cfg/yolov3.cfg yolov3.weights yolov3.conv.81 81 will be created file yolov3.conv.81, then train by using weights file yolov3.conv.81 instead of darknet53.conv.74



> Would it make sense from a ML perspective to push that logic even further and freeze all weights except the yolo layers?

No. [yolo] layer isn't trainable.

**Comment by mmartin56** [comment]
Thanks! Sorry, by 'Yolo layers' I really meant the convolutional layers (with 1x1 filter) right before the yolo layers. So freeze everything but those 3. Would it make sense to do that to train yolo on a new class? You'd think that yolov3.weights, which are trained on coco's 80 classes, would have enough variety in the layer preceding those 1x1 filters to detect pretty much anything?

Thanks again

**Comment by AlexeyAB** [comment]
Since number of this filters= is different, because you use the number of classes is different from 80, then all below weights will be shifted. So you can't use yolov3.weights as pretrained file.


----

In other way - in general:

You can try to set `stopbackward=1` there: https://github.com/AlexeyAB/darknet/blob/bb7d69941cbce4f67f10406395d685ea92be9478/cfg/yolov3.cfg#L598

And set `onlyforward=1` for all [conv] layers after this which you don't want to change.

**Comment by mmartin56** [comment]
Thanks @AlexeyAB
